File: src/core/database/queries/scoped-entry.query.ts
import type { DatabaseContext } from '../connection.js'
import type { ScopedEntry } from '../types.js'

export class ScopedEntryQuery {
  constructor(private ctx: DatabaseContext) {}

  // Get latest version of an entry with all related data in one query
  getLatest(scopeId: number, key: string): ScopedEntry | undefined {
    const row = this.ctx.db
      .prepare(`
        SELECT
          e.id, e.scope_id, e.key, e.created_at as entry_created_at,
          es.is_archived, es.current_version,
          v.version, v.file_path, v.hash, v.description, v.created_at as version_created_at
        FROM entries e
        JOIN entry_status es ON e.id = es.entry_id
        JOIN versions v ON e.id = v.entry_id AND v.version = es.current_version
        WHERE e.scope_id = ? AND e.key = ?
      `)
      .get(scopeId, key) as Record<string, unknown> | undefined

    return row ? this.mapToScopedEntry(row) : undefined
  }

  // Get specific version of an entry
  getByVersion(scopeId: number, key: string, version: number): ScopedEntry | undefined {
    const row = this.ctx.db
      .prepare(`
        SELECT
          e.id, e.scope_id, e.key, e.created_at as entry_created_at,
          es.is_archived,
          v.version, v.file_path, v.hash, v.description, v.created_at as version_created_at
        FROM entries e
        JOIN entry_status es ON e.id = es.entry_id
        JOIN versions v ON e.id = v.entry_id
        WHERE e.scope_id = ? AND e.key = ? AND v.version = ?
      `)
      .get(scopeId, key, version) as Record<string, unknown> | undefined

    return row ? this.mapToScopedEntry(row) : undefined
  }

  // List entries with efficient JOIN
  list(scopeId: number, includeArchived = false, allVersions = false): ScopedEntry[] {
    let query: string

    if (allVersions) {
      query = `
        SELECT
          e.id, e.scope_id, e.key, e.created_at as entry_created_at,
          es.is_archived,
          v.version, v.file_path, v.hash, v.description, v.created_at as version_created_at
        FROM entries e
        JOIN entry_status es ON e.id = es.entry_id
        JOIN versions v ON e.id = v.entry_id
        WHERE e.scope_id = ?
        ${includeArchived ? '' : 'AND es.is_archived = 0'}
        ORDER BY e.key, v.version DESC
      `
    } else {
      query = `
        SELECT
          e.id, e.scope_id, e.key, e.created_at as entry_created_at,
          es.is_archived, es.current_version,
          v.version, v.file_path, v.hash, v.description, v.created_at as version_created_at
        FROM entries e
        JOIN entry_status es ON e.id = es.entry_id
        JOIN versions v ON e.id = v.entry_id AND v.version = es.current_version
        WHERE e.scope_id = ?
        ${includeArchived ? '' : 'AND es.is_archived = 0'}
        ORDER BY e.key
      `
    }

    const rows = this.ctx.db.prepare(query).all(scopeId) as Record<string, unknown>[]
    return rows.map((row) => this.mapToScopedEntry(row))
  }

  // Get all entries for multiple scopes in one query
  listByScopes(scopeIds: number[]): Map<number, ScopedEntry[]> {
    if (scopeIds.length === 0) return new Map()

    const placeholders = scopeIds.map(() => '?').join(',')
    const query = `
      SELECT
        e.id, e.scope_id, e.key, e.created_at as entry_created_at,
        es.is_archived, es.current_version,
        v.version, v.file_path, v.hash, v.description, v.created_at as version_created_at
      FROM entries e
      JOIN entry_status es ON e.id = es.entry_id
      JOIN versions v ON e.id = v.entry_id AND v.version = es.current_version
      WHERE e.scope_id IN (${placeholders})
      AND es.is_archived = 0
      ORDER BY e.scope_id, e.key
    `

    const rows = this.ctx.db.prepare(query).all(...scopeIds) as Record<string, unknown>[]

    const result = new Map<number, ScopedEntry[]>()
    for (const row of rows) {
      const entry = this.mapToScopedEntry(row)
      const scopeEntries = result.get(entry.scopeId) || []
      scopeEntries.push(entry)
      result.set(entry.scopeId, scopeEntries)
    }

    return result
  }

  private mapToScopedEntry(row: Record<string, unknown>): ScopedEntry {
    return {
      id: row.id as number,
      scopeId: row.scope_id as number,
      key: row.key as string,
      version: row.version as number,
      filePath: row.file_path as string,
      hash: row.hash as string,
      description: row.description as string | undefined,
      createdAt: new Date(row.entry_created_at as string),
      isArchived: Boolean(row.is_archived),
    }
  }
}

File: internal/database/scoped_entry_query.go
package database

import (
	"context"
	"database/sql"
	"fmt"

	sqldb "github.com/vault-md/vaultmd/internal/database/sqlc"
)

type ScopedEntryQuery struct {
	ctx *Context
}

func NewScopedEntryQuery(dbCtx *Context) *ScopedEntryQuery {
	return &ScopedEntryQuery{ctx: dbCtx}
}

func (q *ScopedEntryQuery) GetLatest(ctx context.Context, scopeID int64, key string) (*ScopedEntryRecord, error) {
	queries := queriesFromContext(q.ctx)
	if queries == nil {
		return nil, fmt.Errorf("scoped entry query: missing database context")
	}

	row, err := queries.GetScopedEntryLatest(ctx, sqldb.GetScopedEntryLatestParams{ScopeID: scopeID, Key: key})
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}

	record := mapScopedEntryRow(row.EntryID, row.ScopeID, row.Key, row.EntryCreatedAt, row.IsArchived, row.Version, row.FilePath, row.Hash, row.Description)
	return &record, nil
}

func (q *ScopedEntryQuery) GetByVersion(ctx context.Context, scopeID int64, key string, version int64) (*ScopedEntryRecord, error) {
	queries := queriesFromContext(q.ctx)
	if queries == nil {
		return nil, fmt.Errorf("scoped entry query: missing database context")
	}

	row, err := queries.GetScopedEntryByVersion(ctx, sqldb.GetScopedEntryByVersionParams{ScopeID: scopeID, Key: key, Version: version})
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}

	record := mapScopedEntryRow(row.EntryID, row.ScopeID, row.Key, row.EntryCreatedAt, row.IsArchived, row.Version, row.FilePath, row.Hash, row.Description)
	return &record, nil
}

func (q *ScopedEntryQuery) List(ctx context.Context, scopeID int64, includeArchived bool, allVersions bool) ([]ScopedEntryRecord, error) {
	queries := queriesFromContext(q.ctx)
	if queries == nil {
		return nil, fmt.Errorf("scoped entry query: missing database context")
	}

	if allVersions {
		rows, err := queries.ListScopedEntriesAllVersions(ctx, sqldb.ListScopedEntriesAllVersionsParams{ScopeID: scopeID, IncludeArchived: includeArchived})
		if err != nil {
			return nil, err
		}
		result := make([]ScopedEntryRecord, 0, len(rows))
		for _, row := range rows {
			result = append(result, mapScopedEntryRow(row.EntryID, row.ScopeID, row.Key, row.EntryCreatedAt, row.IsArchived, row.Version, row.FilePath, row.Hash, row.Description))
		}
		return result, nil
	}

	rows, err := queries.ListScopedEntriesLatest(ctx, sqldb.ListScopedEntriesLatestParams{ScopeID: scopeID, IncludeArchived: includeArchived})
	if err != nil {
		return nil, err
	}
	result := make([]ScopedEntryRecord, 0, len(rows))
	for _, row := range rows {
		result = append(result, mapScopedEntryRow(row.EntryID, row.ScopeID, row.Key, row.EntryCreatedAt, row.IsArchived, row.Version, row.FilePath, row.Hash, row.Description))
	}
	return result, nil
}

func (q *ScopedEntryQuery) ListByScopes(ctx context.Context, scopeIDs []int64) (map[int64][]ScopedEntryRecord, error) {
	queries := queriesFromContext(q.ctx)
	if queries == nil {
		return nil, fmt.Errorf("scoped entry query: missing database context")
	}

	result := make(map[int64][]ScopedEntryRecord, len(scopeIDs))
	for _, scopeID := range scopeIDs {
		rows, err := queries.ListScopedEntriesLatest(ctx, sqldb.ListScopedEntriesLatestParams{ScopeID: scopeID, IncludeArchived: false})
		if err != nil {
			return nil, err
		}
		entries := make([]ScopedEntryRecord, 0, len(rows))
		for _, row := range rows {
			entries = append(entries, mapScopedEntryRow(row.EntryID, row.ScopeID, row.Key, row.EntryCreatedAt, row.IsArchived, row.Version, row.FilePath, row.Hash, row.Description))
		}
		result[scopeID] = entries
	}
	return result, nil
}

func mapScopedEntryRow(entryID, scopeID int64, key string, entryCreatedAt sql.NullTime, isArchived sql.NullInt64, version int64, filePath, hash string, description sql.NullString) ScopedEntryRecord {
	var descPtr *string
	if description.Valid {
		val := description.String
		descPtr = &val
	}

	return ScopedEntryRecord{
		EntryID:     entryID,
		ScopeID:     scopeID,
		Key:         key,
		Version:     version,
		FilePath:    filePath,
		Hash:        hash,
		Description: descPtr,
		CreatedAt:   optionalTime(entryCreatedAt),
		IsArchived:  optionalBool(isArchived),
	}
}
